src/app_conn_pool.rs

```diff
@@ -0,0 +1,319 @@
+                    app_ws: app_ws.clone(),
+                    opened_at: Timestamp::now(),
+                });
+
+                app_ws
+            }
+        };
+
+        if app_client_lock.len() > self.configuration.max_app_connections as usize {
+            // Find and remove the oldest connection
+            let installed_app_id = app_client_lock
+                .iter()
+                .min_by_key(|(_, v)| v.opened_at)
+                .map(|(k, _)| k.clone())
+                .expect("Invalid lock");
+
+            tracing::warn!(
+                "Reached maximum app connections, removing connection for app: {}",
+                installed_app_id
+            );
+
+            app_client_lock.remove(&installed_app_id);
+        }
+
+        Ok(app_ws)
+    }
+
+    /// Remove an app client from the pool.
+    pub async fn remove_app_client(&self, installed_app_id: &InstalledAppId) {
+        self.app_clients.write().await.remove(installed_app_id);
+    }
+
+    async fn attempt_connect_app_ws(
+        &self,
+        installed_app_id: InstalledAppId,
+        admin_ws: AdminWebsocket,
+        retries: i8,
+    ) -> HcHttpGatewayResult<AppWebsocket> {
+        tracing::debug!(
+            "Attempting to connect to app client for {}",
+            installed_app_id
+        );
+
+        // Get the app port for a compatible app interface, which may be a cached value.
+        let app_port = self
+            .get_app_port(&installed_app_id, admin_ws.clone())
+            .await?;
+        tracing::debug!("Using app port {}", app_port);
+
+        // Issue an app authentication token to allow us to connect a new client.
+        let issued = admin_ws
+            .issue_app_auth_token(IssueAppAuthenticationTokenPayload::for_installed_app_id(
+                installed_app_id.clone(),
+            ))
+            .await?;
+
+        // Presence of host must have been checked to get an admin connection
+        let host = self
+            .configuration
+            .admin_ws_url
+            .host_str()
+            .expect("Must have a host");
+
+        tracing::info!("Connecting to app websocket at {}:{}", host, app_port);
+
+        // Build a connection request
+        let request =
+            ConnectRequest::from(format!("{host}:{app_port}").parse::<SocketAddr>().map_err(
+                |e| {
+                    HcHttpGatewayError::ConfigurationError(format!("Invalid socket address: {}", e))
+                },
+            )?)
+            .try_set_header("Origin", HTTP_GW_ORIGIN)?;
+
+        // Create a websocket client configuration and lower the default timeout. We are connecting
+        // locally to a running Holochain. If requests take longer than 10s then we want to free up
+        // the HTTP gateway to handle other requests.
+        let mut config = WebsocketConfig::CLIENT_DEFAULT;
+        config.default_request_timeout = std::time::Duration::from_secs(10);
```

Comment:
- I assume we've already run `init` on the installed app, before it is called via this gateway (so this will never trigger wasm compliation). If not its too short of a timeout.
- I could imagine heavy zome calls taking longer than 10s, but maybe that's just a hard constraint we're setting for app devs
- Should this timeout be configured in the gateway env vars?

Reply:
10 seconds is indeed quite short. Especially if this gateway is supposed to support remote calls as well, these can easily take several 10s of seconds in some cases.

Reply:
Yes, running `init` is a pre-requisite and the tests have to do that too.
Possible 10s is too low and I agree we should make this a configuration option. Making remote calls is not expected behavior though, and nor is going to the network ideally. This is supposed to serving DHT data and not also fetching it.

https://github.com/holochain/hc-http-gw/pull/17/commits/6d038a3310a574646592df77399e4243bb1c37e4